Documentation for Fine Performance Metrics and Spans (#7945)
dask/distributed: docs/source/fine-performance-metrics.rst

Fine Performance Metrics
========================
.. note::

    This is an experimental feature and may rapidly change without a deprecation cycle.

You may want to investigate where your Dask workload spends the majority of its time;
not only on which tasks, but also *doing what* while running said tasks.

> *Fine performance metrics* are collected automatically by Dask and attempt to answer

Reviewer: Could we first explain what they do before saying what you can use them for?

Author: tweaked

this question.

In order to observe them, you should simply

#. run your workload end-to-end
#. open the Dask dashboard (default for a LocalCluster: `<http://localhost:8787>`_)
#. select ``More...`` -> ``Fine Performance Metrics``

Alternatively, if you're using Jupyter Lab and `dask-labextension
<https://github.com/dask/dask-labextension>`_, you can just drag the
``Fine Performance Metrics`` widget onto your Jupyter dashboard.

.. image:: images/fine-performance-metrics/seconds-full.png

>    :alt: Populated Fine Performance Metrics dashboard

Reviewer: Do you have an example where the function names are readable?

Author: I'm regenerating the pictures after the recent PRs

>    :alt: Populated Fine Performance Metrics dashboard

Reviewer 2: #7962

> The central panel shows, of all the time that the cluster spent doing something, what
> *activities* it spent performing. The most important ones are:

Reviewer 2: nit:
```suggestion
The central panel shows what *activities* the cluster spent its time on. 
The most important ones are:
```

``thread-cpu``

>   CPU time spent by tasks while running on workers

Reviewer: Maybe add a small comment that this is "good" time

Author: done

``thread-noncpu``
  Difference between wall clock time and CPU time spent by tasks while running on
  workers. This is typically either I/O time, CPU contention, or GIL contention.
``idle``
  Time where a worker had a free thread, but nothing to run on it. This is typically
  caused by the workload not being able to fully utilize all threads on the cluster,
  excessive CPU load on the scheduler, or network latency between scheduler and workers.
  This measure does not include time spent while the whole cluster was completely idle.

The grand total of the time shown should roughly add up to the end-to-end runtime of
your workload, multiplied by the number of threads on the cluster.

> The plot on the right shows network transfer time. Note that most of it should be

Reviewer: Maybe use ``The "Send Data, by activity" plot shows network transfer time.`` Giving the plot a label makes it a bit easier to understand for me.

Author: Done

pipelined with task execution, so it may not have an impact. You should worry about this

> only if you have a very large ``idle`` time.

Reviewer: Do you have an example illustrating a worrisome case?

Author: Not at hand. I would need to manually run all tests from coiled/benchmarks that show idle time. Even then, it would be complicated to figure out the cause.
This can be fixed by metering waiting time - but not something we can do now.

> only if you have a very large ``idle`` time.

Author: fixed in https://github.com/dask/distributed/pull/7963

function. There is a filter that allows you to show only selected functions. In the
sample screenshots, you can observe that most of the `thread-noncpu` time is
concentrated - as expected - in functions that are known to be I/O heavy. Here they are
singled out:

.. image:: images/fine-performance-metrics/seconds-IO.png
   :alt: Fine Performance Metrics dashboard, just the I/O functions

And here's everything but those three I/O heavy functions:

.. image:: images/fine-performance-metrics/seconds-not-IO.png
   :alt: Fine Performance Metrics dashboard, except the I/O functions

This tells us an important piece of information: why is `astype`, which is a pure CPU
function, spending so much time occupying workers' threads but without accruing any CPU
time? The answer, almost certainly, it's that it doesn't properly release the GIL.

Fine Performance Metrics don't collect exclusively wall clock timings. We can change the
unit to bytes:

.. image:: images/fine-performance-metrics/bytes.png
   :alt: Fine Performance Metrics dashboard, unit=bytes

The above gives us insights on spill/unspill activity (see :doc:`worker-memory`): in

> this workflow, in 99% of the cases there was enough RAM to accommodate all the data and
> it was unnecessary to retrieve it from disk; in other words we have a 99% *cache hit
> ratio*, which means that we would not get any benefit if we increased RAM, but we would
> likely start seeing slowdowns if we reduced it.

Reviewer 2: This is conflating cache hits and byte hits. Since we're counting bytes here, this is less about the number of times the cache got hit but rather about the amount of data we were able to retrieve from the cache. Maybe this is clearer?

```suggestion
this workflow, 99% of the data was found in memory and it did not have to be retrieved from disk; in other words, we have a 99% *byte hit
ratio*, which means that we would not get much benefit if we increased RAM, but we would
likely start seeing slowdowns if we reduced it.
```

Author: Yes, "bytes" shows cache hit ratio weighted by data size, whereas "count" shows unweighted cache hit ratio e.g. number of accesses. Unless your chunk size is extremely small and you are backed by a disk with very high latency (e.g. an NFS mount), you typically only care about the former.

This said, `cache hit ratio` immediately evokes the correct image in the mind of any reader, whereas `byte hit ratio` definitely doesn't.

Reviewer 2: > This said, `cache hit ratio` immediately evokes the correct image in the mind of any reader, whereas `byte hit ratio` definitely doesn't.

I'd disagree that in the current form it evokes the correct image when thinking about uneven data sizes. For example, if the 1 % of data stem from a single but huge result while we have _many_ small ones, it currently evokes the image that we had to load from cache 1 out of 100 intermediate results. Instead, we had to read only one out of thousands or more. If we want to avoid `byte hit ratio` as a fairly uncommon term, I'd suggest stressing the point that it's weighted by data size. Maybe this is as simple as 
```suggestion
this workflow, only 1% of the data had to be retrieved from disk, whereas 99% of the data could be stored in memory; in other words we have a *cache hit* for 99% of the data, which means that we would not get any benefit if we increased RAM, but we would
likely start seeing slowdowns if we reduced it.
```

Reviewer 2: Either way, this is non-blocking.

Author: tweaked

Reviewer 2: Thanks!

:meth:`~dask.dataframe.DataFrame.map_partitions`), you may want to customize the metrics
they produce. For example, you may want to separate I/O time from `thread-noncpu`:

.. code-block:: python

   from distributed.metrics import context_meter

   @context_meter.meter("I/O")
   def read_some_files():
       ...

   future = client.submit(read_some_files)

In the above example, the wall time spent by the custom function ``read_some_files``
will be logged as "I/O", which is a completely arbitrary activity label.

Alternatively you may want to just label some of the time this way:

.. code-block:: python

   def read_some_files():
       with context_meter.meter("I/O"):
           data = read_from_network(...)
       return preprocess(data)

In the above example, the function is split into an I/O intensive phase,
``read_from_network``, and a CPU-intensive one, ``preprocess``. The
``context_meter.meter`` context manager will subtract the time spent by
``read_from_network`` from ``thread-noncpu``, whereas the time spent by ``preprocess``
will still be a mix of ``thread-cpu`` and ``thread-noncpu`` (the latter may, for
example, highlight GIL contention).

.. note::

>    The ``context_meter.meter`` context manager wraps around code that runs on the

Reviewer: Can we link to these methods? E.g. are there dosctrings available?

Author: done

  activity.
``idle``
  Time where a worker had a free thread, but nothing to run on it. This is typically
  caused by the workload not being able to fully utilize all threads on the cluster,
  network latency between scheduler and workers, or excessive CPU load on the scheduler.
  This measure does not include time spent while the whole cluster was completely idle.
``disk-read``, ``disk-write``, ``compress``, ``decompress``
  Time spent spilling/unspilling to disk due to not having enough memory available.
  See :doc:`worker-memory`.
``executor``, ``offload``, ``other``
  This is overhead from the Dask code and should be typically negligible. However,
  it can be inflated by GIL contention and by spill/unspill activity.

The grand total of the time shown should roughly add up to the end-to-end runtime of
your workload, multiplied by the number of threads on the cluster.

The right panel (**Send data, by activity**) shows network transfer time. Note that most
of it should be pipelined with task execution, so it may not have an impact. You should
worry about this only if you have a very large ``idle`` time.

> The left panel (**Task execution, by function**) shows the same information as the

Reviewer 2: nit: I'd move this above the explanation for the right panel since it picks up on the explanation of the central panel.

Author: done
